docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,416 @@
+bucket.
+
+The simulation can be accessed at:
+[https://raduberinde.github.io/distbucket-v21-06-14/?workload=steps](https://raduberinde.github.io/distbucket-v21-06-14/?workload=steps).
+
+The page shows the granted workload over time for both the prototyped algorithm
+and an ideal (instantly synchronized) token bucket. We also show the graphs of
+total consumption between the two, to see if there is any systematic (long-term)
+divergence.
+
+The workload can be changed from the dropdown or the yaml can be manually
+edited. Each workload function is a sum of terms; the supported terms can be
+inferred from the
+[code](https://github.com/RaduBerinde/raduberinde.github.io/blob/master/distbucket-v21-06-14/lib/data.go#L77).
+
+The code is available here:
+[https://github.com/RaduBerinde/raduberinde.github.io/blob/master/distbucket-v21-06-14/lib/dist_token_bucket_3.go](https://github.com/RaduBerinde/raduberinde.github.io/blob/master/distbucket-v21-06-14/lib/dist_token_bucket_3.go).
+
+## System table
+
+The system table uses a "ledger" approach; each change to the state of a global
+tenant bucket appends a row to the table. Each row has a unique operation UUID
+and a sequence number which is unique within that tenant. The operation UUID
+allows detection of duplicate requests, to ensure idempotency. The sequence
+numbers allow locating the latest entry for each tenant.
+
+Schema for the system table:
+```sql
+CREATE TABLE tenant_usage (
+  tenant_id INT NOT NULL,
+  seq INT NOT NULL,
+
+  op_id UUID NOT NULL,  -- TODO: maybe INT?
+  op_time TIMESTAMP NOT NULL,
+
+  ru_burst_limit FLOAT NOT NULL,
+  ru_rate FLOAT NOT NULL,
+
+  ru_current FLOAT NOT NULL,
+  current_share_sum FLOAT NOT NULL,
+
+  -- Cumulative usage statistics.
+  total_ru_usage FLOAT NOT NULL,
+  total_read_requests INT NOT NULL,
+  total_read_bytes INT NOT NULL,
+  total_write_requests INT NOT NULL,
+  total_write_bytes INT NOT NULL,
+  total_sql_pod_cpu_seconds FLOAT NOT NULL, -- TODO: Maybe milliseconds and INT8?
+
+  PRIMARY KEY (tenant_id, seq DESC),
+  UNIQUE (tenant_id, op_id) -- could also be UNIQUE(op_id)
+);
+```
+
+An operation to the bucket is implemented with a transaction:
+```sql
+-- Generate random op_id.
+BEGIN;
+-- Get the latest state (with the largest sequence number).
+SELECT FOR UPDATE * FROM tenant_usage WHERE tenant_id=.. ORDER BY seq DESC LIMIT 1;
+-- Calculate new state.
+-- Set the new state, using the next sequence number.
+INSERT INTO tenant_usage (tenant_id, seq, ...) (tenant_id, seq+1, ...);
+END;
+```
+
+Any failed request is retried with the same `op_id`. If the previous request
+went through, the transaction will error out with a conflict on `op_id`,
+allowing special handling of this case.
+
+We perform periodic garbage collection to keep the table from growing. Each
+operation can trigger garbage collection (with random chance).
+
+```sql
+-- by number of entries
+DELETE FROM tenant_usage WHERE tenant_id=.. AND seq < (
```

Comment:
ABP (always be paginating) here and below

Reply:
This query will perform poorly because of the queue-like workload. Either at the beginning or end it will end up scanning through the tombstone desert. It would be more effective to make GC stateful (remembering the largest seqno since GC'ed).